### PR TITLE
Use the ItemNormalizer in the correct context

### PR DESCRIPTION
**Pull request by soyuka**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | n/a will prevent some
| License       | MIT
| Doc PR        | na

This PR prevents the use of ApiPlatform's ItemNormalizer when not in the correct context. 
This means that:
- if we're using Symfony without any supported formats and outside ApiPlatform context, we're using symfony normalizers
- if it's a hypermedia format (`jsonld`, `hal`, `jsonapi`) we're using ApiPlatform's normalizers even if we're inside symfony
- if it's another format but we're using ApiPlatform we're still using our ItemNormalizers

### CODE REVIEW

File: src/Serializer/ItemNormalizer.php [hunk 1]
```diff
@@ -98,4 +104,12 @@ private function updateObjectToPopulate(array $data, array &$context)
             $context[self::OBJECT_TO_POPULATE] = $this->iriConverter->getItemFromIri(sprintf('%s/%s', $this->iriConverter->getIriFromResourceClass($context['resource_class']), $data[$identifier]), $context + ['fetch_data' => true]);
         }
     }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function supportsNormalization($data, $format = null)
+    {
+        return (($context['use_api_platform'] ?? false) || \in_array($format, self::FORMATS, true)) && parent::supportsNormalization($data, $format);
```
Comment (dunglas) [inline src/Serializer/ItemNormalizer.php]: Can you introduce a const for `use_api_platform`?

**Comment by alanpoulain** [comment]
Could you add a description to explain why you needed to do this, for future reference?

File: src/Serializer/ItemNormalizer.php [hunk 2]
```diff
@@ -98,4 +104,12 @@ private function updateObjectToPopulate(array $data, array &$context)
             $context[self::OBJECT_TO_POPULATE] = $this->iriConverter->getItemFromIri(sprintf('%s/%s', $this->iriConverter->getIriFromResourceClass($context['resource_class']), $data[$identifier]), $context + ['fetch_data' => true]);
         }
     }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function supportsNormalization($data, $format = null)
```
Comment (dunglas) [inline src/Serializer/ItemNormalizer.php]: `$context` isn't defined, you must implements the `ContextAwareInterface`

File: src/Serializer/AbstractItemNormalizer.php [hunk 3]
```diff
@@ -144,7 +144,7 @@ public function normalize($object, $format = null, array $context = [])
      */
     public function supportsDenormalization($data, $type, $format = null)
     {
-        if (ItemNormalizer::FORMAT === $format) {
+        if (ElasticsearchItemNormalizer::FORMAT === $format) {
```
Comment (dunglas) [inline src/Serializer/AbstractItemNormalizer.php]: Why this change?

Comment (soyuka) [reply src/Serializer/AbstractItemNormalizer.php]: Because it's easier to read, it can be confusing if you read `ItemNormalizer`

**Comment by soyuka** [comment]
Currently failing because the `$context` never reaches the `supportsNormalization`, maybe because of https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Serializer/Serializer.php#L240

**Comment by dunglas** [comment]
@alanpoulain to prevent errors like this one: https://github.com/api-platform/core/issues/2563

A better strategy would be to use another instance of the `Serializer` class for API Platform than the Symfony built-in one. But it's not doable as is, we need to introduce some kind of factories to wire a fully featured Serializer in Symfony first. This context option is a quick fix.

File: src/Serializer/AbstractItemNormalizer.php [hunk 4]
```diff
@@ -93,8 +101,12 @@ public function __construct(PropertyNameCollectionFactoryInterface $propertyName
     /**
      * {@inheritdoc}
      */
-    public function supportsNormalization($data, $format = null)
+    public function supportsNormalization($data, $format = null, array $context = [])
     {
+        if (!(($context[static::USE_API_PLATFORM] ?? false) || \in_array($format, self::FORMATS, true))) {
```
Comment (antograssiot) [inline src/Serializer/AbstractItemNormalizer.php]: just asking, but given that this method now relies on the context to answer, can we still systematically return `true` in the `hasCacheableSupportsMethod` ?

Comment (teohhanhui) [reply src/Serializer/AbstractItemNormalizer.php]: We can't. And that's one of the reasons why this fix cannot work...

Comment (soyuka) [reply src/Serializer/AbstractItemNormalizer.php]: how should we fix it ? :P

Comment (teohhanhui) [reply src/Serializer/AbstractItemNormalizer.php]: Revert 😝

Comment (teohhanhui) [reply src/Serializer/AbstractItemNormalizer.php]: Sometimes we gotta acknowledge we've made a wrong turn...

Comment (dunglas) [reply src/Serializer/AbstractItemNormalizer.php]: Just return `false` in `hasCacheableSupportsMethod`.
@teohhanhui revert what? All our new input/output system? It doesn't look like a good idea. It's a very good move, and a long awaited one.

The proper fix for this specific problem is to use our own instance of the serializer instead of the global one provided by Symfony. To do so, we need to patch SF first, to provide convenient factories to wire and configure a serializer (basically, `FrameworkExtension` must be revamped). It will also benefit to Doctrine JSON ODM, and many other libraries.
But it's a hard work, and we need a quick fix first, such as this one. We'll improve later.

**Comment by teohhanhui** [comment]
Uhh... Do you want me to stop working on the alternative solution?
